src/Access/ReplicatedAccessStorage.cpp

```diff
@@ -0,0 +1,657 @@
+{
+    LOG_DEBUG(getLogger(), "Updating entity {}", toString(id));
+    Notifications notifications;
+    SCOPE_EXIT({ notify(notifications); });
+
+    std::lock_guard lock{mutex};
+    for (size_t retry=0; retry<10; ++retry)
+    {
+        if (tryUpdateNoLock(id, update_func, notifications))
+            return;
+    }
+    throw Exception(ErrorCodes::UNEXPECTED_ZOOKEEPER_ERROR, "Cannot update access entity");
+}
+
+
+bool ReplicatedAccessStorage::tryUpdateNoLock(const UUID & id, const UpdateFunc & update_func, Notifications & notifications)
+{
+    auto zookeeper = get_zookeeper();
+    const String entity_uuid = toString(id);
+    const String entity_path = zookeeper_path + "/uuid/" + entity_uuid;
+
+    String old_entity_definition;
+    Coordination::Stat stat;
+    const auto uuid_exists = zookeeper->tryGet(entity_path, old_entity_definition, &stat);
+    if (!uuid_exists)
+        throwNotFound(id);
+
+    const AccessEntityPtr old_entity = parseAccessEntity(old_entity_definition, "znode", entity_path);
+    const AccessEntityPtr new_entity = update_func(old_entity);
+
+    if (!new_entity->isTypeOf(old_entity->getType()))
+        throwBadCast(id, new_entity->getType(), new_entity->getName(), old_entity->getType());
+
+    const EntityTypeInfo type_info = EntityTypeInfo::get(new_entity->getType());
+
+    Coordination::Requests ops;
+    const String new_entity_definition = serializeAccessEntity(*new_entity);
+    ops.emplace_back(zkutil::makeSetRequest(entity_path, new_entity_definition, stat.version));
+
+    const String & old_name = old_entity->getName();
+    const String & new_name = new_entity->getName();
+    if (new_name != old_name)
+    {
+        auto old_name_path = zookeeper_path + "/" + type_info.unique_char + "/" + escapeForFileName(old_name);
+        auto new_name_path = zookeeper_path + "/" + type_info.unique_char + "/" + escapeForFileName(new_name);
+        ops.emplace_back(zkutil::makeRemoveRequest(old_name_path, -1));
+        ops.emplace_back(zkutil::makeCreateRequest(new_name_path, entity_uuid, zkutil::CreateMode::Persistent));
+    }
+
+    Coordination::Responses responses;
+    const auto res = zookeeper->tryMulti(ops, responses);
+    if (res == Coordination::Error::ZNODEEXISTS)
+    {
+        throwNameCollisionCannotRename(new_entity->getType(), old_name, new_name);
+    }
+    else if (res == Coordination::Error::ZNONODE)
+    {
+        throwNotFound(id);
+    }
+    else if (res == Coordination::Error::ZBADVERSION)
+    {
+        return false; // Retry
+    }
+    else if (res != Coordination::Error::ZOK)
+    {
+        /// Other codes are unexpected, will throw
+        zkutil::KeeperMultiException::check(res, ops, responses);
+        assert(false);
+        __builtin_unreachable();
+    }
+
+    refreshEntityNoLock(zookeeper, id, notifications);
+    return true;
+}
+
+
+void ReplicatedAccessStorage::runMainThread()
```

Comment:
I don't think the name is good, this is not the main thread, this is a worker thread or a refreshing thread.